### ISSUE TITLE
New Lint: unused_pure_return_value

### ISSUE THREAD

**Issue by PhilosophicalProgrammer** [rust-lang/rust-clippy#6910]
### What it does

It checks for function detected to be pure. I know rust doesn't have pure functions, but Clippy could make a best guess at whether or not a function is pure in the CS sense (no side effects, same return value for the same arguments). In this case, a call to it with an unused return value should be warned to do nothing. This should probably be warn by default, as the proposed implementation is quite paranoid about functions being impure (all mutable dereferencing from arguments is considered "impure", as well as interior mutability) and is very unlikely to give false positives without some inherent design flaw.

### Drawbacks

Maybe perf?

### Example

```rust
impure();
purefn();
impure();
```

Could be written as this code with reduced complexity:

```rust
impure();
impure();
```

### Implementation

Check a function and the tree of functions it calls. If any of the calls in the call tree perform an impure operation in any branch, then the function and the functions that call it are marked as impure. Impure operations include dereferencing a mutable reference from an argument, dereferencing a mutable raw pointer, any sort of assembly, using statics, intrinsics, and calling FFI functions. Recursive calls would be ignored.

This would cover all I/O functions. For example, if I have this function, then the `println!` calls an impure function because it calls FFI for the printing, which makes the function itself impure, and so this lint wouldn't apply to it:

```rust
fn this_is_impure() {
    println!("Hello, world!");
}
```

This is also impure because of a possible panic in a branch, even though it's obviously not possible (evaluating branches is not in the scope of this lint):

```rust
fn this_is_impure() {
     Ok(()).unwrap();
}
```

However, a common example of a function that would be pure by this method of detection would be the "new" function here:

```rust
impl SomeType {
    fn new() -> Self {
        Self {
            value: pure_value_creator(),
        }
    }
}
```

**Comment by mikerite** [comment]
Isn't this covered by the [must_use_candidate](https://rust-lang.github.io/rust-clippy/master/index.html#must_use_candidate)?
